validate/vmspec.go

```diff
@@ -0,0 +1,275 @@
+	var buf [8]byte
+	_, err := io.ReadFull(vm.code, buf[:])
+	if err != nil {
+		return 0, err
+	}
+	return binary.LittleEndian.Uint64(buf[:]), nil
+}
+
+func (vm *mockSpecVM) opdSize() uint32 {
+	return uint32(len(vm.opdStack))
+}
+
+func (vm *mockSpecVM) ctrlSize() uint32 {
+	return uint32(len(vm.ctrlStack))
+}
+
+func (vm *mockSpecVM) pushOpd(Type wasm.ValueType) error {
+	if Type != wasm.ValueType(wasm.BlockTypeEmpty) {
+		vm.opdStack = append(vm.opdStack, Type)
+	}
+	return nil
+}
+
+func (vm *mockSpecVM) popOpd() (wasm.ValueType, error) {
+	topFrame, err := vm.topCtrl()
+	if err != nil {
+		return ValueTypeUnk, err
+	}
+
+	if topFrame.height == vm.opdSize() && topFrame.unreachable {
+		return ValueTypeUnk, nil
+	} else if topFrame.height == vm.opdSize() {
+		return ValueTypeUnk, errors.New("operand stack underflow")
+	}
+
+	if len(vm.opdStack) == 0 {
+		return ValueTypeUnk, errors.New("code0 logic error")
+	}
+
+	Type := vm.opdStack[len(vm.opdStack)-1]
+
+	vm.opdStack = vm.opdStack[:len(vm.opdStack)-1]
+	return Type, nil
+}
+
+func (vm *mockSpecVM) topCtrl() (*ctrlFrame, error) {
+	if len(vm.ctrlStack) == 0 {
+		return nil, errors.New("no block frame in stack")
+	}
+
+	return &vm.ctrlStack[len(vm.ctrlStack)-1], nil
+}
+
+func (vm *mockSpecVM) pickCtrl(index uint32) (*ctrlFrame, error) {
+	if len(vm.ctrlStack) == 0 {
+		return nil, errors.New("no block frame in stack")
+	}
+
+	/*always use + other then sub to compare on uint32.*/
+	if uint32(len(vm.ctrlStack)) < index+1 {
+		return nil, errors.New("block frame execeed max")
+	}
+
+	return &vm.ctrlStack[uint32(len(vm.ctrlStack))-index-1], nil
+}
+
+func (vm *mockSpecVM) popOpdExpect(expect wasm.ValueType) (wasm.ValueType, error) {
+	if expect == wasm.ValueType(wasm.BlockTypeEmpty) {
+		return wasm.ValueType(wasm.BlockTypeEmpty), nil
+	}
+
+	actual, err := vm.popOpd()
+	if err != nil {
+		return ValueTypeUnk, err
+	}
+
+	if actual == ValueTypeUnk {
+		return expect, nil
+	}
+
+	if expect == ValueTypeUnk {
+		return actual, nil
+	}
+
+	if actual != expect {
+		return ValueTypeUnk, errors.New("not expect type")
+	}
+
+	return actual, nil
+}
+
+func (vm *mockSpecVM) pushCtrl(labelTypes wasm.ValueType, endType wasm.ValueType, fType frameType) error {
+	frame := ctrlFrame{
+		labelTypes:  labelTypes,
+		endType:     endType,
+		height:      vm.opdSize(),
+		unreachable: false,
+		fType:       fType,
+	}
+
+	vm.ctrlStack = append(vm.ctrlStack, frame)
+	return nil
+}
+
+func (vm *mockSpecVM) popCtrl() (wasm.ValueType, error) {
+	topFrame, err := vm.topCtrl()
+	if err != nil {
+		return ValueTypeUnk, err
+	}
+
+	_, err = vm.popOpdExpect(topFrame.endType)
+	if err != nil {
+		return ValueTypeUnk, err
+	}
+
+	if vm.opdSize() != topFrame.height {
+		return ValueTypeUnk, errors.New("stack overflow")
+	}
+
+	vm.ctrlStack = vm.ctrlStack[:len(vm.ctrlStack)-1]
+	return topFrame.endType, nil
+}
+
+func (vm *mockSpecVM) unreachable() error {
+	topFrame, err := vm.topCtrl()
+	if err != nil {
+		return nil
```

Comment:
why return `nil`?, and it is possible to encounter this error since `unreachable` opcode must be in some block already?

Reply:
it's a mistake. should be return err. yes, at least in function block frame.